### PR TITLE
Feature/numa domains

### LINKED ISSUE #3170: Revisit NUMA aware scheduling
Check current state for Tianle's use case.

### PR DESCRIPTION
**Pull request by andre-merzky**
This expands the app level scheduling with NUMA awareness.  Fixes #3170

### CODE REVIEW

File: src/radical/pilot/pilot.py [hunk 1]
```diff
@@ -343,8 +343,19 @@ def nodelist(self):
...
+            node_list       = resource_details.get('node_list')
+
+            if not node_list:
+                return None
+
+            # only create NUMA resources if a numa domain map is available
+            if not numa_domain_map:
+                nodes = [NodeResources(node) for node in node_list]
+            else:
+                nodes = [NumaNodeResources(node, numa_domain_map)
```
Comment (mtitov) [inline src/radical/pilot/pilot.py:356]: `NumaNodeResources` - missing its import

File: src/radical/pilot/agent/resource_manager/base.py [hunk 2]
```diff
@@ -216,17 +236,18 @@ def init_from_scratch(self):
...
+        rm_info.cores_per_node        = self._cfg.cores_per_node
+        rm_info.gpus_per_node         = self._cfg.gpus_per_node
+        rm_info.lfs_per_node          = self._cfg.lfs_size_per_node
+        rm_info.lfs_path              = ru.expand_env(self._cfg.lfs_path_per_node)
 
         rm_info.threads_per_gpu  = 1
         rm_info.mem_per_gpu      = None
-        rm_info.mem_per_node     = self._rcfg.mem_per_node or 0
+        rm_info.mem_per_node     = self._rcfg.mem_per_node    or 0
+        rm_info.numa_domain_map  = self._rcfg.numa_domain_map or {}
```
Comment (mtitov) [inline src/radical/pilot/agent/resource_manager/base.py:250]: we don't activate numa-domains yet, right? (i.e., Agent doesn't use this information)

File: src/radical/pilot/agent/resource_manager/base.py [hunk 3]
```diff
@@ -40,39 +40,59 @@ class RMInfo(ru.TypedDict):
...
-            'lfs_path'         : str,           # node local FS path
-            'mem_per_node'     : int,           # memory per node (MB)
+            'lfs_per_node'         : int,           # node local FS size (MB)
+            'lfs_path'             : str,           # node local FS path
+            'mem_per_node'         : int,           # memory per node (MB)
 
-            'details'          : {None: None},  # dict of launch method info
-            'launch_methods'   : {str: None},   # dict of launch method info
+            'details'              : {None: None},  # dict of launch method info
+            'lm_info'              : {str : None},  # dict of launch method info
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/agent/resource_manager/base.py:64]: Thanks, cleaned up.

File: src/radical/pilot/agent/resource_manager/base.py [hunk 4]
```diff
@@ -40,39 +40,59 @@ class RMInfo(ru.TypedDict):
...
+            'gpus_per_node'        : 0,
+            'threads_per_gpu'      : 1,
+            'mem_per_gpu'          : 0,
+
+            'lfs_per_node'         : 0,
+            'lfs_path'             : '/tmp/',
+            'mem_per_node'         : 0,
+
+            'details'              : dict(),
+            'lm_info'              : dict(),
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/agent/resource_manager/base.py:92]: same

File: src/radical/pilot/agent/resource_manager/base.py (hunk 2 above)
Comment (andre-merzky) [reply src/radical/pilot/agent/resource_manager/base.py:250]: That is correct - but the upcoming scheduler changes will use it, so I wanted to have that readily available.

File: src/radical/pilot/pilot.py (hunk 1 above)
Comment (andre-merzky) [reply src/radical/pilot/pilot.py:356]: Thanks, fixed.

File: src/radical/pilot/resource_config.py [hunk 5]
```diff
@@ -39,8 +39,10 @@
 MEM_PER_NODE           = 'mem_per_node'
 CORES_PER_NODE         = 'cores_per_node'
 GPUS_PER_NODE          = 'gpus_per_node'
+NUMA_DOMAIN_MAP        = 'numa_domain_map'
 SYSTEM_ARCHITECTURE    = 'system_architecture'
 SCATTERED              = 'scattered'
+SERVICES               = 'services'
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/resource_config.py:45]: Probably - I removed that.

File: src/radical/pilot/resource_config.py [hunk 6]
```diff
@@ -136,8 +138,10 @@ class ResourceConfig(ru.TypedDict):
         MEM_PER_NODE           : int         ,
         CORES_PER_NODE         : int         ,
         GPUS_PER_NODE          : int         ,
+        NUMA_DOMAIN_MAP        : {int: None} ,
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/resource_config.py:141]: Yes - json jas a problem with int keys, but the zmq serialization (msgpack) does not have a problem with that.

File: src/radical/pilot/resource_config.py [hunk 7]
```diff
@@ -754,5 +836,70 @@ def release_slots(self, slots: List[Slot]) -> None:
...
+            self.__domains__   = list()
+            self.__n_domains__ = len(numa_domain_map)
+
+            for domain_id, domain_descr in numa_domain_map.items():
+
+                n = NodeResources(from_dict)
+                n.index = n.index
+                n.name  = '%s.%s' % (self.name, domain_id)
+                n.cores = [_RO(index=i) for i in domain_descr.cores]
+                n.gpus  = [_RO(index=i) for i in domain_descr.gpus]
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/resource_config.py:879]: Thanks.

File: src/radical/pilot/resource_config.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mtitov) [reply src/radical/pilot/resource_config.py:45]: It does look like a leftover
- we read `services` from PD
https://github.com/radical-cybertools/radical.pilot/blob/3dad2d16faa2df909aaa6f8c8feff9cc573d45c4/src/radical/pilot/pmgr/launching/base.py#L563
- then we reassign `services` with info from the resource config
https://github.com/radical-cybertools/radical.pilot/blob/3dad2d16faa2df909aaa6f8c8feff9cc573d45c4/src/radical/pilot/pmgr/launching/base.py#L594


and then we write this information into the agent config, not a resource config

[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/resource_config.py]: Yep, you are right - thanks, is fixed now.

[the remark this answers is not in the document]
Comment (mtitov) [reply src/radical/pilot/resource_config.py]: Thank you!

**Review by mtitov — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**